ts/query.go

```diff
@@ -0,0 +1,431 @@
+	// Advance real iterators until nextReal has offset >= the interpolated
+	// offset.
+	for ii.nextReal.valid && ii.nextReal.offset < ii.offset {
+		ii.prevReal = ii.nextReal
+		ii.nextReal.advance()
+	}
+}
+
+// isValid returns true if this interpolatingIterator still points to valid data.
+func (ii *interpolatingIterator) isValid() bool {
+	return ii.nextReal.valid
+}
+
+// avg returns the average value at the current offset for this iterator.
+func (ii *interpolatingIterator) avg() float64 {
+	if !ii.isValid() {
+		return 0
+	}
+	if ii.nextReal.offset == ii.offset {
+		return ii.nextReal.sample().Average()
+	}
+	// Cannot interpolate if previous value is invalid.
+	if !ii.prevReal.valid {
+		return 0
+	}
+
+	// Linear interpolation of value at the current offset.
+	off := float64(ii.offset)
+	nextAvg := ii.nextReal.sample().Average()
+	nextOff := float64(ii.nextReal.offset)
+	prevAvg := ii.prevReal.sample().Average()
+	prevOff := float64(ii.prevReal.offset)
+	return prevAvg + (nextAvg-prevAvg)*(off-prevOff)/(nextOff-prevOff)
+}
+
+// newIterator returns an interpolating iterator for the given dataSpan. The
+// iterator is initialized to offset 0.
+func (ds *dataSpan) newIterator() interpolatingIterator {
+	if len(ds.datas) == 0 {
+		return interpolatingIterator{}
+	}
+
+	// The first data index necessarily contains the positive offset closest to
+	// 0. Use a binary search to find the lowest positive offset (or the zero
+	// offset).
+	data := ds.datas[0]
+	innerIdx := sort.Search(len(data.Samples), func(i int) bool {
+		return data.offsetAt(i) >= 0
+	})
+
+	iterator := interpolatingIterator{
+		offset: 0,
+		nextReal: dataSpanIterator{
+			dataSpan:  ds,
+			dataIdx:   0,
+			sampleIdx: innerIdx,
+			offset:    data.offsetAt(innerIdx),
+			valid:     true,
+		},
+	}
+
+	// If innerIdx > 0, then we can compute a "previous" iterator as well; this
+	// will let us interpolate a 0 value if it is not actually present in the
+	// data.
+	if innerIdx > 0 {
+		iterator.prevReal = dataSpanIterator{
+			dataSpan:  ds,
+			dataIdx:   0,
+			sampleIdx: innerIdx - 1,
+			offset:    data.offsetAt(innerIdx - 1),
+			valid:     true,
+		}
+	}
+
+	return iterator
+}
+
+// iteratorSet is used to coordinate the advancement of multiple
+// interpolatingIterators. It is also used to aggregate the current values of
+// iterators in the set.
+//
+// All valid iterators in the set will have the same offset at all times. During
+// advancement, the next offset is chosen by finding the individual iterator
+// with the lowest value of nextReal.offset; in other words, the iteratorSet
+// will visit each possible offset in sequence, skipping offsets for which *no*
+// iterators have real data.  If even a single iterator has real data at an
+// offset, that offset will eventually be visited.
+//
+// In order to faciliate finding the lowest value of nextReal.offset, the set is
+// organized as a min heap using Go's heap package.
+type iteratorSet []interpolatingIterator
```

Comment:
(mere suggestion):
I understand that `interpolatingIteratorSetInterator` is not a great name, but after understanding what this is all about I would have preferred the name `unionIterator`, because that's basically what it does: It iterates over the union of offsets (with real values) of the contained `interpolatingIterator`s.

> // A unionIterator jointly advances multiple interpolatingIterators, visiting precisely those offsets for which at least one of the underlying interpolating iterators has a real (that is, non-interpolated) value.